NetBSD mmap()

Previously reported to the Monero project as Issue [#6592](https://github.com/monero-project/monero/issues/6592):

Compiling on NetBSD 9 (x86_64) fails with this message:

    /home/t/usr/src/RandomX/src/virtual_memory.cpp: In function ‘void* allocLargePagesMemory(std::size_t)’:
    /home/t/usr/src/RandomX/src/virtual_memory.cpp:147:83: error: ‘MAP_HUGETLB’ was not declared in this scope
      mem = mmap(nullptr, bytes, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_HUGETLB | MAP_POPULATE, -1, 0);
                                                                                       ^~~~~~~~~~~

Compiling succeeds if the following is added immediately below the "//OpenBSD does not support huge pages" comment in src/virtualmemory.cpp

    #elif defined(__NetBSD__)
        #define RESERVED_FLAGS PROT_MPROTECT(PROT_EXEC)
        mem = mmap(nullptr, bytes, PROT_READ | PROT_WRITE | RESERVED_FLAGS, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);

(and also a ` && !defined(__NetBSD__)` is appended to the end of the `#elif !defined(__OpenBSD__)` line in src/tests/affinity.cpp). This is a quick cut-and-paste from elsewhere, since I don't know C++, and I'm not suggesting it's sensible.

Then, running randomx-tests gives the following output:
    [ 1] Cache initialization                     ... PASSED
    [ 2] SuperscalarHash generator                ... PASSED
    [ 3] randomx_reciprocal                       ... PASSED
    [ 4] randomx_reciprocal_fast                  ... PASSED
    [ 5] Dataset initialization (interpreter)     ... PASSED
    [ 6] Dataset initialization (compiler)        ... terminate called after throwing an instance of 'std::runtime_error'
      what():  mprotect failed
    Abort (core dumped)

I refreshed #187 and tried again.

randomx_benchmark is still OK. Here's the last few lines of output from `gdb ./randomx-tests`:

    [89] Cache initialization: SSSE3              ... PASSED
    [90] Cache initialization: AVX2               ... SKIPPED
    [91] Hash batch test                          ... assertion "equalsHex(hash1, "639183aae1bf4c9a35884cb46b09cad9175f04efd7684e7262a0ac1c2f0b4e3f")" failed: file "/home/t/usr/src/RandomX/src/tests/tests.cpp", line 1072, function "main()::<lambda()>"
    
    Program received signal SIGABRT, Aborted.
    0x00007bd92ff678aa in _lwp_kill () from /usr/lib/libc.so.12
    (gdb) bt
    #0  0x00007bd92ff678aa in _lwp_kill () from /usr/lib/libc.so.12
    #1  0x00007bd92ff6715a in abort () at /usr/src/lib/libc/stdlib/abort.c:74
    #2  0x00007bd92fea84d7 in __assert13 (file=<optimized out>, 
        line=<optimized out>, function=<optimized out>, failedexpr=<optimized out>)
        at /usr/src/lib/libc/gen/assert.c:72
    #3  0x0000000000406eb3 in void runTest<main::{lambda()#86}>(char const*, bool, main::{lambda()#86}) [clone .constprop.107] ()
    #4  0x0000000000415a25 in main ()
    (gdb) quit
    A debugging session is active.
    
    	Inferior 1 [process 4255] will be killed.
    
    Quit anyway? (y or n) y

I also checked out #189, but it didn't compile. I wonder if you've accidentally put the wrong src/virtual_memory.cpp in #189; since I don't see any "defined(__NetBSD__)" in there and, it's falling through to the not-defined-anything case:

    [ 26%] Building CXX object CMakeFiles/randomx.dir/src/virtual_memory.cpp.o
    /home/t/usr/src/RandomX/src/virtual_memory.cpp: In function ‘void* allocLargePagesMemory(std::size_t)’:
    /home/t/usr/src/RandomX/src/virtual_memory.cpp:147:83: error: ‘MAP_HUGETLB’ was not declared in this scope
      mem = mmap(nullptr, bytes, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_HUGETLB | MAP_POPULATE, -1, 0);
                                                                                       ^~~~~~~~~~~

--- a/src/jit_compiler.hpp
+++ b/src/jit_compiler.hpp
@@ -35,3 +35,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #else
 #include "jit_compiler_fallback.hpp"
 #endif
+
+#if defined(__OpenBSD__) || defined(__NetBSD__)
+#define RANDOMX_FORCE_SECURE
+#endif

--- a/src/randomx.cpp
+++ b/src/randomx.cpp
@@ -43,7 +43,7 @@ extern "C" {
 	randomx_flags randomx_get_flags() {
 		randomx_flags flags = RANDOMX_HAVE_COMPILER ? RANDOMX_FLAG_JIT : RANDOMX_FLAG_DEFAULT;
 		randomx::Cpu cpu;
-#ifdef __OpenBSD__
+#ifdef RANDOMX_FORCE_SECURE
 		if (flags == RANDOMX_FLAG_JIT) {
 			flags |= RANDOMX_FLAG_SECURE;
 		}

--- a/src/tests/affinity.cpp
+++ b/src/tests/affinity.cpp
@@ -65,7 +65,7 @@ set_thread_affinity(std::thread::native_handle_type thread,
             (thread_policy_t)&policy, 1);
 #elif defined(_WIN32) || defined(__CYGWIN__)
     rc = SetThreadAffinityMask(reinterpret_cast<HANDLE>(thread), 1ULL << cpuid) == 0 ? -2 : 0;
-#elif !defined(__OpenBSD__) && !defined(__FreeBSD__) && !defined(__ANDROID__)
+#elif !defined(__OpenBSD__) && !defined(__FreeBSD__) && !defined(__ANDROID__) && !defined(__NetBSD__)
     cpu_set_t cs;
     CPU_ZERO(&cs);
     CPU_SET(cpuid, &cs);

--- a/src/tests/benchmark.cpp
+++ b/src/tests/benchmark.cpp
@@ -40,6 +40,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "../dataset.hpp"
 #include "../blake2/endian.h"
 #include "../common.hpp"
+#include "../jit_compiler.hpp"
 #ifdef _WIN32
 #include <windows.h>
 #include <VersionHelpers.h>
@@ -199,7 +200,7 @@ int main(int argc, char** argv) {
 		}
 		if (jit) {
 			flags |= RANDOMX_FLAG_JIT;
-#ifdef __OpenBSD__
+#ifdef RANDOMX_FORCE_SECURE
 			flags |= RANDOMX_FLAG_SECURE;
 #endif
 		}
@@ -211,7 +212,7 @@ int main(int argc, char** argv) {
 	if (miningMode) {
 		flags |= RANDOMX_FLAG_FULL_MEM;
 	}
-#ifndef __OpenBSD__
+#ifndef RANDOMX_FORCE_SECURE
 	if (secure) {
 		flags |= RANDOMX_FLAG_SECURE;
 	}

--- a/src/tests/tests.cpp
+++ b/src/tests/tests.cpp
@@ -143,7 +143,7 @@ int main() {
 		randomx::JitCompiler jit;
 		jit.generateSuperscalarHash(cache->programs, cache->reciprocalCache);
 		jit.generateDatasetInitCode();
-#ifdef __OpenBSD__
+#ifdef RANDOMX_FORCE_SECURE
 		jit.enableExecution();
 #else
 		jit.enableAll();
@@ -954,7 +954,7 @@ int main() {
 		assert(ibc.memMask == randomx::ScratchpadL3Mask);
 	});
 
-#ifdef __OpenBSD__
+#ifdef RANDOMX_FORCE_SECURE
 	vm = randomx_create_vm(RANDOMX_FLAG_DEFAULT | RANDOMX_FLAG_SECURE, cache, nullptr);
 #else
 	vm = randomx_create_vm(RANDOMX_FLAG_DEFAULT, cache, nullptr);
@@ -1009,10 +1009,10 @@ int main() {
 		vm = nullptr;
 		cache = randomx_alloc_cache(RANDOMX_FLAG_JIT);
 		initCache("test key 000");
-#ifdef __OpenBSD__
-		vm = randomx_create_vm(RANDOMX_FLAG_DEFAULT | RANDOMX_FLAG_SECURE, cache, nullptr);
+#ifdef RANDOMX_FORCE_SECURE
+		vm = randomx_create_vm(RANDOMX_FLAG_JIT | RANDOMX_FLAG_SECURE, cache, nullptr);
 #else
-		vm = randomx_create_vm(RANDOMX_FLAG_DEFAULT, cache, nullptr);
+		vm = randomx_create_vm(RANDOMX_FLAG_JIT, cache, nullptr);
 #endif
 	}
 

--- a/src/virtual_memory.cpp
+++ b/src/virtual_memory.cpp
@@ -94,7 +94,12 @@ void* allocMemoryPages(std::size_t bytes) {
 	if (mem == nullptr)
 		throw std::runtime_error(getErrorMessage("allocMemoryPages - VirtualAlloc"));
 #else
-	mem = mmap(nullptr, bytes, PAGE_READWRITE, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
+	#if defined(__NetBSD__)
+		#define RESERVED_FLAGS PROT_MPROTECT(PROT_EXEC)
+	#else
+		#define RESERVED_FLAGS 0
+	#endif
+	mem = mmap(nullptr, bytes, PAGE_READWRITE | RESERVED_FLAGS, MAP_ANONYMOUS | MAP_PRIVATE, -1, 0);
 	if (mem == MAP_FAILED)
 		throw std::runtime_error("allocMemoryPages - mmap failed");
 #endif
@@ -141,7 +146,7 @@ void* allocLargePagesMemory(std::size_t bytes) {
 	mem = mmap(nullptr, bytes, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, VM_FLAGS_SUPERPAGE_SIZE_2MB, 0);
 #elif defined(__FreeBSD__)
 	mem = mmap(nullptr, bytes, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_ALIGNED_SUPER, -1, 0);
-#elif defined(__OpenBSD__)
+#elif defined(__OpenBSD__) || defined(__NetBSD__)
 	mem = MAP_FAILED; // OpenBSD does not support huge pages
 #else
 	mem = mmap(nullptr, bytes, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_HUGETLB | MAP_POPULATE, -1, 0);

